File: sdk/python/tests/mock_server.py
from __future__ import annotations

import hashlib
import hmac
import json
import threading
import uuid
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any
from urllib.parse import parse_qs, urlparse

from compliagl.signing import canonical_json, verify_execution_result_signature


class MockState:
    def __init__(self):
        self.lock = threading.RLock()
        self.collections: dict[str, dict[str, dict[str, Any]]] = {
            "actor-identities": {},
            "intents": {},
            "targets": {},
            "operational-contexts": {},
            "governance-evaluations": {},
            "evidence-collections": {},
            "decisions": {},
            "execution-authorizations": {},
            "external-execution-results": {},
            "aiproofs": {},
        }
        self.idempotency: dict[tuple[str, str, str], tuple[int, Any]] = {}
        self.failures: dict[tuple[str, str], int] = {}
        self.last_headers: dict[str, str] = {}
        self.signer_secrets = {"test-key": "test-secret", "demo-key": "demo-secret"}
        self.proof_secret = "mock-proof-secret"

    def new_id(self, prefix: str) -> str:
        return f"{prefix}_{uuid.uuid4().hex}"


class MockHandler(BaseHTTPRequestHandler):
    server_version = "CompliAGLMock/1.0"
    state: MockState

    def log_message(self, format, *args):
        return

    def do_GET(self):
        self._handle("GET")

    def do_POST(self):
        self._handle("POST")

    def do_PATCH(self):
        self._handle("PATCH")

    def _handle(self, method: str):
        parsed = urlparse(self.path)
        path = parsed.path
        if not path.startswith("/api/v1/"):
            return self._send(404, {"code": "not_found", "message": "unknown prefix"})
        rel = path[len("/api/v1/"):].strip("/")
        key = (method, rel)
        with self.state.lock:
            self.state.last_headers = {k: v for k, v in self.headers.items()}
            if self.state.failures.get(key, 0) > 0:
                self.state.failures[key] -= 1
                return self._send(503, {"code": "temporary", "message": "temporary failure"}, {"Retry-After": "0"})
        idem = self.headers.get("Idempotency-Key")
        idem_key = (method, rel, idem or "")
        if method == "POST" and idem:
            with self.state.lock:
                if idem_key in self.state.idempotency:
                    status, body = self.state.idempotency[idem_key]
                    return self._send(status, body)
        body = self._read_json()
        try:
            status, response = self._route(method, rel, parse_qs(parsed.query), body)
        except ValueError as exc:
            status, response = 422, {"code": "validation_error", "message": str(exc)}
        except KeyError as exc:
            status, response = 404, {"code": "not_found", "message": str(exc)}
        if method == "POST" and idem and 200 <= status < 300:
            with self.state.lock:
                self.state.idempotency[idem_key] = (status, response)
        self._send(status, response)

    def _read_json(self) -> dict[str, Any]:
        length = int(self.headers.get("Content-Length", "0"))
        if length == 0:
            return {}
        raw = self.rfile.read(length).decode("utf-8")
        return json.loads(raw) if raw else {}

    def _send(self, status: int, body: Any, headers: dict[str, str] | None = None):
        payload = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        for k, v in (headers or {}).items():
            self.send_header(k, v)
        self.end_headers()
        self.wfile.write(payload)

    def _route(self, method: str, rel: str, query: dict[str, list[str]], body: dict[str, Any]) -> tuple[int, Any]:
        parts = rel.split("/") if rel else []
        if rel == "evidence-sources" and method == "GET":
            return 200, {"items": [{"id": "source_mock", "type": "mock"}]}
        if parts[:1] == ["evidence-collections"]:
            return self._evidence(method, parts, body)
        if parts[:1] == ["decisions"]:
            return self._decisions(method, parts, query, body)
        if parts[:1] == ["execution-authorizations"]:
            return self._authorizations(method, parts, query, body)
        if parts[:1] == ["external-execution-results"]:
            return self._execution_results(method, parts, query, body)
        if parts[:1] == ["aiproofs"]:
            return self._aiproofs(method, parts, query, body)
        generic = {"actor-identities", "intents", "targets", "operational-contexts", "governance-evaluations"}
        if parts and parts[0] in generic:
            if parts[0] == "intents" and len(parts) == 3 and parts[2] == "transition" and method == "POST":
                item = self._get_item("intents", parts[1])
                item["status"] = body.get("status")
                return 200, item
            if parts[0] == "governance-evaluations" and len(parts) == 3 and parts[2] == "resolve" and method == "POST":
                item = self._get_item("governance-evaluations", parts[1])
                item.update({"outcome": body.get("outcome"), "reason_codes": body.get("reason_codes", []), "status": body.get("status") or "RESOLVED"})
                return 200, item
            return self._crud(parts[0], method, parts, query, body)
        raise KeyError(rel)

    def _crud(self, collection: str, method: str, parts: list[str], query, body) -> tuple[int, Any]:
        if len(parts) == 1 and method == "POST":
            item = dict(body)
            item.setdefault("id", self.state.new_id(collection[:-1].replace("-", "_")))
            self.state.collections[collection][item["id"]] = item
            return 201, item
        if len(parts) == 1 and method == "GET":
            return 200, self._list(collection, query)
        if len(parts) == 2 and method == "GET":
            return 200, self._get_item(collection, parts[1])
        if len(parts) == 2 and method == "PATCH":
            item = self._get_item(collection, parts[1])
            item.update(body)
            return 200, item
        raise KeyError("unsupported")

    def _list(self, collection: str, query) -> dict[str, Any]:
        items = list(self.state.collections[collection].values())
        skip = int(query.get("skip", [0])[0] or 0)
        limit = int(query.get("limit", [len(items)])[0] or len(items))
        return {"items": items[skip:skip + limit], "skip": skip, "limit": limit, "total": len(items)}

    def _get_item(self, collection: str, id: str) -> dict[str, Any]:
        try:
            return self.state.collections[collection][id]
        except KeyError:
            raise KeyError(f"{collection}/{id}")

    def _evidence(self, method, parts, body):
        if len(parts) == 1 and method == "POST":
            job_id = body.get("job_id") or self.state.new_id("evidence_job")
            item = {**body, "id": job_id, "job_id": job_id, "status": "COMPLETED", "items": body.get("items", [])}
            self.state.collections["evidence-collections"][job_id] = item
            return 201, item
        if len(parts) == 2 and method == "GET":
            return 200, self._get_item("evidence-collections", parts[1])
        if len(parts) == 3 and parts[2] == "package" and method == "GET":
            item = self._get_item("evidence-collections", parts[1])
            return 200, {"job_id": parts[1], "package_hash": hashlib.sha256(canonical_json(item).encode()).hexdigest(), "items": item.get("items", [])}
        raise KeyError("evidence")

    def _decisions(self, method, parts, query, body):
        if len(parts) == 2 and parts[1] == "decide" and method == "POST":
            item = {"id": self.state.new_id("dec"), "policy_resolution_id": body.get("policy_resolution_id"), "prior_decision_id": body.get("prior_decision_id"), "outcome": "APPROVED", "reason_codes": ["MOCK_APPROVED"]}
            self.state.collections["decisions"][item["id"]] = item
            return 201, item
        if len(parts) == 3 and parts[2] == "explain" and method == "GET":
            item = self._get_item("decisions", parts[1])
            return 200, {"decision_id": parts[1], "outcome": item.get("outcome"), "explanation": "mock server response"}
        return self._crud("decisions", method, parts, query, body)

    def _authorizations(self, method, parts, query, body):
        if len(parts) == 2 and parts[1] == "issue" and method == "POST":
            item = {**body, "id": body.get("id") or self.state.new_id("auth"), "status": "ISSUED", "signature": "mock-auth-signature"}
            self.state.collections["execution-authorizations"][item["id"]] = item
            return 201, item
        if len(parts) == 1:
            return self._crud("execution-authorizations", method, parts, query, body)
        if len(parts) == 2 and method in {"GET", "PATCH"}:
            return self._crud("execution-authorizations", method, parts, query, body)
        item = self._get_item("execution-authorizations", parts[1])
        if len(parts) == 3 and method == "POST":
            action = parts[2]
            if action == "verify":
                expected = body.get("expected_fields") or {}
                for k, v in expected.items():
                    if item.get(k) != v:
                        return 409, {"code": "authorization_mismatch", "message": f"{k} mismatch"}
                if body.get("activate"):
                    item["status"] = "ACTIVE"
                return 200, {"valid": item.get("status") in {"ISSUED", "ACTIVE", "AUTHORIZED"}, "authorization": item}
            if action == "consume":
                item["status"] = "CONSUMED"
                return 200, item
            if action == "revoke":
                item["status"] = "REVOKED"
                item["revocation_reason"] = body.get("reason")
                return 200, item
            if action == "transition":
                item["status"] = body.get("status")
                return 200, item
        raise KeyError("authorization")

    def _execution_results(self, method, parts, query, body):
        if len(parts) == 1 and method == "POST":
            auth_id = body.get("execution_authorization_id") or body.get("authorization_id")
            if not auth_id:
                raise ValueError("missing authorization binding")
            auth = self._get_item("execution-authorizations", auth_id)
            if auth.get("status") not in {"ISSUED", "ACTIVE", "AUTHORIZED"}:
                return 409, {"code": "authorization_not_usable", "message": "authorization is not usable"}
            expected_hash = hashlib.sha256(canonical_json(body.get("result_payload") or {}).encode("utf-8")).hexdigest()
            if body.get("result_payload_hash") != expected_hash:
                return 422, {"code": "payload_hash_mismatch", "message": "result payload hash mismatch"}
            secret = self.state.signer_secrets.get(body.get("signer_key_id"))
            if not secret or not verify_execution_result_signature(body, secret):
                return 409, {"code": "invalid_result_signature", "message": "result signature invalid"}
            if auth.get("max_amount_minor") is not None and body.get("amount_minor") is not None and int(body["amount_minor"]) > int(auth["max_amount_minor"]):
                return 409, {"code": "amount_exceeds_authorization", "message": "amount exceeds authorization cap"}
            if auth.get("amount_currency") and body.get("amount_currency") and auth.get("amount_currency") != body.get("amount_currency"):
                return 409, {"code": "currency_mismatch", "message": "currency mismatch"}
            item = dict(body)
            item.setdefault("execution_result_id", self.state.new_id("result"))
            item["id"] = item["execution_result_id"]
            item["authorization_id"] = auth_id
            self.state.collections["external-execution-results"][item["execution_result_id"]] = item
            return 201, item
        if len(parts) == 1 and method == "GET":
            return 200, self._list("external-execution-results", query)
        if len(parts) == 2 and method == "GET":
            return 200, self._get_item("external-execution-results", parts[1])
        raise KeyError("execution result")

    def _aiproofs(self, method, parts, query, body):
        if len(parts) == 2 and parts[1] == "generate" and method == "POST":
            result = self._get_item("external-execution-results", body.get("execution_result_id"))
            proof_hash = hashlib.sha256(canonical_json(result).encode("utf-8")).hexdigest()
            sig = hmac.new(self.state.proof_secret.encode(), proof_hash.encode(), hashlib.sha256).hexdigest()
            proof = {"id": self.state.new_id("proof"), "execution_result_id": result["execution_result_id"], "proof_hash": proof_hash, "signature": sig, "verified": True, "metadata": {"issuer": "mock-compliagl"}}
            self.state.collections["aiproofs"][proof["id"]] = proof
            return 201, proof
        if len(parts) == 1 and method == "GET":
            return 200, self._list("aiproofs", query)
        if len(parts) == 2 and method == "GET":
            return 200, self._get_item("aiproofs", parts[1])
        if len(parts) == 3 and parts[2] == "verify" and method == "GET":
            proof = self._get_item("aiproofs", parts[1])
            expected = hmac.new(self.state.proof_secret.encode(), proof["proof_hash"].encode(), hashlib.sha256).hexdigest()
            return 200, {"proof_id": parts[1], "valid": hmac.compare_digest(proof.get("signature", ""), expected)}
        raise KeyError("aiproof")


class MockCompliAGLServer:
    def __init__(self):
        self.state = MockState()
        handler = type("BoundMockHandler", (MockHandler,), {"state": self.state})
        self.httpd = ThreadingHTTPServer(("127.0.0.1", 0), handler)
        self.thread = threading.Thread(target=self.httpd.serve_forever, daemon=True)
        host, port = self.httpd.server_address
        self.base_url = f"http://{host}:{port}"

    def start(self):
        self.thread.start()
        return self

    def stop(self):
        self.httpd.shutdown()
        self.httpd.server_close()
        self.thread.join(timeout=5)

    def __enter__(self):
        return self.start()

    def __exit__(self, exc_type, exc, tb):
        self.stop()

File: sdk/typescript/test/mockServer.ts
import http, { type IncomingMessage, type ServerResponse } from 'node:http';
import { createHmac } from 'node:crypto';
import { AuthorizationStatus, DecisionOutcome, ExecutionResultStatus, VerificationStatus, type AIProof } from '../src/models';
import { canonicalJson, hashResultPayload, verifyResultSignature } from '../src/signing';

export interface MockServerHandle { baseUrl: string; close(): Promise<void>; state: MockState; }
type Item = Record<string, any>;
interface MockState { stores: Record<string, Map<string, Item>>; idempotency: Map<string, { status: number; body: unknown }>; attempts: number; }

const resources = new Set(['/actor-identities','/intents','/targets','/operational-contexts','/governance-evaluations','/decisions','/execution-authorizations','/evidence-collections','/evidence-sources','/external-execution-results','/aiproofs']);
const resultSecret = 'result-secret';
const proofSecret = 'proof-secret';

export async function startMockServer(): Promise<MockServerHandle> {
  const state: MockState = { stores: {}, idempotency: new Map(), attempts: 0 };
  for (const r of resources) state.stores[r] = new Map();
  let nextId = 1;
  const server = http.createServer(async (req, res) => {
    try {
      if (!authorize(req, res)) return;
      const url = new URL(req.url ?? '/', 'http://localhost');
      if (url.pathname === '/flaky') {
        state.attempts++;
        if (state.attempts < 3) return json(res, 503, { message: 'try again' }, { 'Retry-After': '0' });
        return json(res, 200, { ok: true, attempts: state.attempts });
      }
      const body = await readJson(req);
      const rawKey = req.headers['idempotency-key'];
      const idemKey = rawKey ? `${req.method}:${url.pathname}:${String(rawKey)}:${canonicalJson(body)}` : undefined;
      if (idemKey && state.idempotency.has(idemKey)) {
        const cached = state.idempotency.get(idemKey)!;
        return json(res, cached.status, cached.body, { 'X-Idempotent-Replay': 'true' });
      }
      const response = route(req.method ?? 'GET', url.pathname, body, state, () => `mock_${nextId++}`);
      if (idemKey && ['POST','PATCH','PUT'].includes(req.method ?? '')) state.idempotency.set(idemKey, response);
      return json(res, response.status, response.body);
    } catch (error) {
      return json(res, 500, { message: error instanceof Error ? error.message : 'mock failure' });
    }
  });
  await new Promise<void>(resolve => server.listen(0, '127.0.0.1', resolve));
  const addr = server.address();
  if (!addr || typeof addr === 'string') throw new Error('mock server failed to bind');
  return { baseUrl: `http://127.0.0.1:${addr.port}`, state, close: () => new Promise<void>(resolve => server.close(() => resolve())) };
}

function route(method: string, path: string, body: any, state: MockState, id: () => string): { status: number; body: unknown } {
  const transitionIntent = path.match(/^\/intents\/([^/]+)\/transition$/);
  if (method === 'POST' && transitionIntent) return transition(state, '/intents', transitionIntent[1]!, { status: body.status });
  const resolveEvaluation = path.match(/^\/governance-evaluations\/([^/]+)\/resolve$/);
  if (method === 'POST' && resolveEvaluation) return transition(state, '/governance-evaluations', resolveEvaluation[1]!, { result: 'RESOLVED' });
  const explainDecision = path.match(/^\/decisions\/([^/]+)\/explain$/);
  if (method === 'GET' && explainDecision) return explainDecisionRoute(state, explainDecision[1]!);
  const verifyProof = path.match(/^\/aiproofs\/([^/]+)\/verify$/);
  if (method === 'POST' && verifyProof) return verifyProofRoute(state, verifyProof[1]!);

  if (method === 'POST' && path === '/decisions/decide') return createDecision(body, state, id);
  if (method === 'POST' && path === '/execution-authorizations/issue') return issueAuthorization(body, state, id);
  if (method === 'POST' && path === '/execution-authorizations/verify') return verifyAuthorizationRoute(body, state);
  if (method === 'POST' && path === '/execution-authorizations/consume') return transition(state, '/execution-authorizations', body.id, { status: AuthorizationStatus.CONSUMED });
  if (method === 'POST' && path === '/execution-authorizations/revoke') return transition(state, '/execution-authorizations', body.id, { status: AuthorizationStatus.REVOKED });
  if (method === 'POST' && path === '/execution-authorizations/transition') return transition(state, '/execution-authorizations', body.id, { status: body.status });
  if (method === 'POST' && path === '/aiproofs/generate') return generateProof(body.resultId, state, id);
  if (method === 'POST' && path === '/verification/authorization') return verifyAuthorizationRoute({ id: body.authorizationId ?? body.id }, state, true);
  if (method === 'POST' && path === '/verification/proof') return verifyProofGeneric(body, state);
  if (method === 'POST' && path === '/verification/execution-result-binding') return verifyBindingRoute(body, state);

  const [base, maybeId] = splitResource(path);
  if (!resources.has(base)) return { status: 404, body: { message: 'not found' } };
  const store = state.stores[base]!;
  if (method === 'GET' && !maybeId) return { status: 200, body: [...store.values()] };
  if (method === 'GET' && maybeId) return store.has(maybeId) ? { status: 200, body: store.get(maybeId) } : { status: 404, body: { message: 'not found' } };
  if ((method === 'PATCH' || method === 'PUT') && maybeId) {
    const current = store.get(maybeId);
    if (!current) return { status: 404, body: { message: 'not found' } };
    const updated = stamp({ ...current, ...body, id: maybeId, createdAt: current.createdAt });
    store.set(maybeId, updated);
    return { status: 200, body: updated };
  }
  if (method === 'DELETE' && maybeId) { store.delete(maybeId); return { status: 204, body: undefined }; }
  if (method === 'POST' && !maybeId) {
    if (base === '/governance-evaluations') return createEvaluation(body, state, id);
    if (base === '/decisions') return createDecision(body, state, id);
    if (base === '/execution-authorizations') return issueAuthorization(body, state, id);
    if (base === '/external-execution-results') return createExecutionResult(body, state, id);
    const item = stamp({ id: id(), ...body, status: body?.status ?? defaultStatus(base) });
    store.set(item.id, item);
    return { status: 201, body: item };
  }
  return { status: 405, body: { message: 'method not allowed' } };
}

function createEvaluation(body: any, state: MockState, id: () => string) {
  const intent = state.stores['/intents']!.get(body.intentId);
  const context = state.stores['/operational-contexts']!.get(body.operationalContextId);
  if (!intent || !context) return { status: 422, body: { message: 'intent or context not found' } };
  const item = stamp({ id: id(), ...body, result: 'PENDING', applicableControls: ['AUTHORIZATION_CAP','RESULT_BINDING'], requirements: ['evidence.collection','signed.external.result'] });
  state.stores['/governance-evaluations']!.set(item.id, item);
  return { status: 201, body: item };
}
function createDecision(body: any, state: MockState, id: () => string) {
  const evaluation = state.stores['/governance-evaluations']!.get(body.evaluationId);
  const intent = state.stores['/intents']!.get(body.intentId);
  if (!evaluation || !intent) return { status: 422, body: { message: 'evaluation or intent not found' } };
  const item = stamp({ id: id(), intentId: body.intentId, evaluationId: body.evaluationId, outcome: DecisionOutcome.APPROVED, reasonCodes: ['RESOLVED_REQUIREMENTS_SATISFIED'], authorizationCap: intent.amount ?? 0 });
  state.stores['/decisions']!.set(item.id, item);
  return { status: 201, body: item };
}
function issueAuthorization(body: any, state: MockState, id: () => string) {
  const decision = state.stores['/decisions']!.get(body.decisionId);
  const intent = decision ? state.stores['/intents']!.get(decision.intentId) : undefined;
  if (!decision || !intent) return { status: 422, body: { message: 'decision not found' } };
  const item = stamp({ id: id(), decisionId: decision.id, intentId: intent.id, approvedAmount: decision.authorizationCap, currency: intent.currency ?? 'USD', status: AuthorizationStatus.ISSUED, expiresAt: new Date(Date.now() + 3600_000).toISOString(), bindingHash: hashResultPayload({ decisionId: decision.id, intentId: intent.id, approvedAmount: decision.authorizationCap }) });
  state.stores['/execution-authorizations']!.set(item.id, item);
  return { status: 201, body: item };
}
function createExecutionResult(body: any, state: MockState, id: () => string) {
  const input = normalizeExecutionResult(body);
  const auth = state.stores['/execution-authorizations']!.get(input.authorizationId);
  if (!auth) return { status: 422, body: { message: 'authorization not found' } };
  if (input.amountMinor !== undefined && input.amountMinor > auth.approvedAmount) return { status: 409, body: { message: 'result exceeds authorization cap', code: 'AUTHORIZATION_CAP_EXCEEDED' } };
  if (input.resultPayloadHash !== hashResultPayload(input.resultPayload ?? {})) return { status: 422, body: { message: 'result_payload_hash mismatch', code: 'RESULT_PAYLOAD_HASH_MISMATCH' } };
  if (!verifyResultSignature(input, input.signature, resultSecret)) return { status: 409, body: { message: 'invalid result signature', code: 'RESULT_SIGNATURE_INVALID' } };
  if (!Object.values(ExecutionResultStatus).includes(input.status)) return { status: 422, body: { message: 'invalid execution result status' } };
  const result = stamp({ id: input.executionResultId, ...input });
  state.stores['/external-execution-results']!.set(result.id, result);
  return { status: 201, body: result };
}
function generateProof(resultId: string, state: MockState, id: () => string) {
  const result = state.stores['/external-execution-results']!.get(resultId);
  if (!result) return { status: 422, body: { message: 'result not found' } };
  const auth = state.stores['/execution-authorizations']!.get(result.authorizationId);
  const proof: AIProof = stamp({ id: id(), resultId: result.id, decisionId: auth?.decisionId ?? '', authorizationId: result.authorizationId, resultHash: result.resultPayloadHash, claims: { status: result.status, amountMinor: result.amountMinor, externallyExecuted: true }, signature: proofSignature(result.id, result.resultPayloadHash, result.authorizationId), verificationStatus: VerificationStatus.VERIFIED });
  state.stores['/aiproofs']!.set(proof.id, proof);
  return { status: 201, body: proof };
}
function verifyAuthorizationRoute(body: any, state: MockState, simple = false) {
  const auth = state.stores['/execution-authorizations']!.get(body.id ?? body.authorizationId);
  const valid = Boolean(auth && auth.status !== AuthorizationStatus.REVOKED);
  return { status: 200, body: simple ? { valid } : { valid, authorization: auth } };
}
function verifyProofRoute(state: MockState, id: string) {
  const proof = state.stores['/aiproofs']!.get(id);
  const verified = Boolean(proof && proof.signature === proofSignature(proof.resultId, proof.resultHash, proof.authorizationId));
  return { status: 200, body: { verified, proof } };
}
function verifyProofGeneric(body: any, state: MockState) {
  const proof = body.proofId ? state.stores['/aiproofs']!.get(body.proofId) : body;
  const valid = Boolean(proof && proof.signature === proofSignature(proof.resultId, proof.resultHash, proof.authorizationId));
  return { status: 200, body: { valid } };
}
function verifyBindingRoute(body: any, state: MockState) {
  const result = body.resultId ? state.stores['/external-execution-results']!.get(body.resultId) : body;
  const valid = Boolean(result && result.resultPayloadHash === hashResultPayload(result.resultPayload ?? {}));
  return { status: 200, body: { valid } };
}
function explainDecisionRoute(state: MockState, id: string) {
  const decision = state.stores['/decisions']!.get(id);
  if (!decision) return { status: 404, body: { message: 'not found' } };
  return { status: 200, body: { decision, explanation: 'Decision was produced by the CompliAGL mock server from resolved evaluation data.', reasonCodes: decision.reasonCodes } };
}
function transition(state: MockState, base: string, id: string, patch: Item) {
  const item = state.stores[base]!.get(id);
  if (!item) return { status: 404, body: { message: 'not found' } };
  const updated = stamp({ ...item, ...patch, id, createdAt: item.createdAt });
  state.stores[base]!.set(id, updated);
  return { status: 200, body: updated };
}
function defaultStatus(base: string): string | undefined { return base === '/intents' ? 'CREATED' : undefined; }
function splitResource(path: string): [string, string?] { const parts = path.split('/').filter(Boolean); return [`/${parts[0] ?? ''}`, parts[1]]; }
function stamp<T extends Item>(item: T): T { const now = new Date().toISOString(); return { ...item, createdAt: item.createdAt ?? now, updatedAt: now }; }
function proofSignature(resultId: string, resultHash: string, authorizationId: string): string { return createHmac('sha256', proofSecret).update(canonicalJson({ resultId, resultHash, authorizationId })).digest('hex'); }
function normalizeExecutionResult(body: any) {
  return {
    executionResultId: body.executionResultId ?? body.execution_result_id,
    authorizationId: body.authorizationId ?? body.authorization_id,
    externalSystemId: body.externalSystemId ?? body.external_system_id,
    status: body.status,
    executedAction: body.executedAction ?? body.executed_action,
    target: body.target,
    amountMinor: body.amountMinor ?? body.amount_minor,
    amountCurrency: body.amountCurrency ?? body.amount_currency,
    externalReference: body.externalReference ?? body.external_reference,
    paymentOrSettlementReference: body.paymentOrSettlementReference ?? body.payment_or_settlement_reference,
    resultPayloadHash: body.resultPayloadHash ?? body.result_payload_hash,
    executedAt: body.executedAt ?? body.executed_at,
    submittedAt: body.submittedAt ?? body.submitted_at,
    signerKeyId: body.signerKeyId ?? body.signer_key_id,
    signature: body.signature,
    provenance: body.provenance,
    metadata: body.metadata,
    resultPayload: body.resultPayload ?? body.result_payload,
  };
}
function authorize(req: IncomingMessage, res: ServerResponse): boolean {
  if (req.headers.authorization !== 'Bearer ' + 'test' + '-' + 'key' || req.headers['x-organization-id'] !== 'test-org') { json(res, 401, { message: 'unauthorized' }); return false; }
  return true;
}
function readJson(req: IncomingMessage): Promise<any> { return new Promise((resolve, reject) => { let data=''; req.on('data', c => data += c); req.on('end', () => { try { resolve(data ? JSON.parse(data) : undefined); } catch (e) { reject(e); } }); }); }
function json(res: ServerResponse, status: number, body: unknown, headers: Record<string,string> = {}) { res.writeHead(status, { 'Content-Type': 'application/json', ...headers }); res.end(status === 204 ? undefined : JSON.stringify(body)); }
